File: articles/john-von-neumann-and-the-architecture-of-modern-computing/haskell/FetchExecute.hs
module Main where

step :: Int -> (String, Int) -> Int
step _ ("LOAD", x) = x
step acc ("ADD", x) = acc + x
step acc _ = acc

main :: IO ()
main = do
  let program = [("LOAD", 2), ("ADD", 3), ("STORE", 0), ("HALT", 0)]
  let acc = foldl step 0 program
  putStrLn ("accumulator=" ++ show acc)

File: articles/john-von-neumann-and-the-architecture-of-modern-computing/prolog/fetch_execute.pl
step(acc(_), load(X), acc(X)).
step(acc(A), add(X), acc(B)) :- B is A + X.
step(acc(A), store(Address), acc(A)) :- format("store address=~w value=~w~n", [Address, A]).
step(acc(A), halt, acc(A)) :- format("halt accumulator=~w~n", [A]).

% Query example:
% ?- step(acc(0), load(2), S1), step(S1, add(3), S2), step(S2, store(0), S3), step(S3, halt, _).
